src/callback_registry_table.h

```diff
@@ -48,41 +68,108 @@ class CallbackRegistryTable {
+      shared_ptr<CallbackRegistry> parent = getRegistry(parent_id);
       if (parent == nullptr) {
         Rcpp::stop("Can't create registry. Parent with id %d does not exist.", parent_id);
       }
       registry->parent = parent;
       parent->children.push_back(registry);
     }
 
-    registries[id] = shared_ptr<CallbackRegistry>(registry);
+    // Would be better to use .emplace() to avoid copy-constructor, but that
+    // requires C++11.
+    registries[id] = RegistryHandle(registry, true);
   }
 
   // Returns a shared_ptr to the registry. If the registry is not present in
   // the table, or if the target CallbackRegistry has already been deleted,
   // then the shared_ptr is empty.
-  shared_ptr<CallbackRegistry> get(int id) {
+  shared_ptr<CallbackRegistry> getRegistry(int id) {
     Guard guard(&mutex);
     if (!exists(id)) {
       return shared_ptr<CallbackRegistry>();
     }
     // If the target of the shared_ptr has been deleted, then this is an empty
     // shared_ptr.
-    return registries[id];
+    return registries[id].registry;
+  }
+
+  uint64_t scheduleCallback(void (*func)(void*), void* data, double delaySecs, int loop_id) {
+    // This method can be called from any thread
+    Guard guard(&mutex);
+
+    shared_ptr<CallbackRegistry> registry = getRegistry(loop_id);
+    if (registry == nullptr) {
+      return 0;
+    }
+    return doExecLater(registry, func, data, delaySecs, true);
+  }
+
+  // This is called when the R loop handle referring to a CallbackRegistry is
+  // destroyed. Returns true if the CallbackRegistry exists and this function
+  // has not previously been called on it; false otherwise.
+  bool notifyRRefDeleted(int id) {
+    ASSERT_MAIN_THREAD()
+    Guard guard(&mutex);
+
+    if (!exists(id)) {
+      return false;
+    }
+
+    if (registries[id].r_ref_exists) {
+      registries[id].r_ref_exists = false;
+      this->pruneRegistries();
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  // Iterate over all registries, and remove a registry when:
+  // * If the loop has a parent:
+  //   * There's no R loop object referring to it, AND the registry is empty.
+  // * If the loop does not have a parent:
+  //   * There's no R loop object referring to it. (Dont' need the registry to
+  //     be empty, because if there's no parent and no R reference to the loop,
+  //     there is no way to execute callbacks in the registry.)
+  void pruneRegistries() {
+    ASSERT_MAIN_THREAD()
+    Guard guard(&mutex);
+
+    std::map<int, RegistryHandle>::iterator it = registries.begin();
+
+    // Iterate over all registries. Remove under the following conditions:
+    // * There are no more R loop handle references to the registry, AND
+    // * The registry is empty, OR the registry has no parent.
+    // This logic is equivalent to the logic describing the function, just in
+    // a different order.
+    while (it != registries.end()) {
```

Comment:
Please add a comment indicating that parents always come earlier than their children in the iteration order, so the case of a parent being pruned leading to children/grandchildren being pruned will work within a single iteration.